service/debugger/debugger.go

```diff
@@ -442,45 +443,80 @@ func (d *Debugger) CreateBreakpoint(requestedBp *api.Breakpoint) (*api.Breakpoin
 				}
 			}
 		}
-		addr, err = proc.FindFileLocation(d.target, fileName, requestedBp.Line)
+		addrs, err = proc.FindFileLocation(d.target, fileName, requestedBp.Line)
 	case len(requestedBp.FunctionName) > 0:
-		addr, err = proc.FindFunctionLocation(d.target, requestedBp.FunctionName, requestedBp.Line)
+		addrs, err = proc.FindFunctionLocation(d.target, requestedBp.FunctionName, requestedBp.Line)
 	default:
-		addr = requestedBp.Addr
+		addrs = []uint64{requestedBp.Addr}
 	}
 
 	if err != nil {
 		return nil, err
 	}
 
-	bp, err := d.target.SetBreakpoint(addr, proc.UserBreakpoint, nil)
+	createdBp, err := createLogicalBreakpoint(d.target, addrs, requestedBp)
 	if err != nil {
 		return nil, err
 	}
-	if err := copyBreakpointInfo(bp, requestedBp); err != nil {
-		if _, err1 := d.target.ClearBreakpoint(bp.Addr); err1 != nil {
-			err = fmt.Errorf("error while creating breakpoint: %v, additionally the breakpoint could not be properly rolled back: %v", err, err1)
+	d.log.Infof("created breakpoint: %#v", createdBp)
+	return createdBp, nil
+}
+
+// createLogicalBreakpoint creates one physical breakpoint for each address
+// ins addrs and associates all of them with the same logical breakpoint.
+func createLogicalBreakpoint(p proc.Process, addrs []uint64, requestedBp *api.Breakpoint) (*api.Breakpoint, error) {
+	bps := make([]*proc.Breakpoint, len(addrs))
+	var err error
+	for i := range addrs {
+		var err2 error
```

Comment:
Why have err and err2 here? We assign `err = err2` on error and break and check `err` anyways.